crates/mmr/src/tree.rs

```diff
@@ -0,0 +1,123 @@
+//! Generic binary Merkle tree implementation built on `MerkleHasher`.
+//!
+//! This complements the MMR by providing standard “binary Merkle tree”
+//! functionality: construct from leaves, generate inclusion proofs, and verify
+//! those proofs.
+
+use crate::hasher::MerkleHasher;
+use crate::proof::MerkleProof;
+
+/// Simple binary Merkle tree backed by in-memory levels.
+///
+/// Construction duplicates the last node when a level has odd length.
+#[derive(Clone, Debug)]
+pub struct BinaryMerkleTree<MH: MerkleHasher> {
+    /// Level 0 is leaves; last level has a single root.
+    levels: Vec<Vec<MH::Hash>>,
```

Comment:
This requires a bunch of extra allocs.

The trick here is to flatten out the merkle tree into a single vec of length 2n-1, starting with the leafs and then appending each pair to the end, going up the tree.  So for 8 leafs the vec looks like:

```
a b c d e f g h ab cd ef gh abcd efgh abcdefgh
```

This really benefits from using bitshifting tricks to figure out where things go, but the key observation is that doing `level_local_offset >> 1` gives you the level local offset in the next level.